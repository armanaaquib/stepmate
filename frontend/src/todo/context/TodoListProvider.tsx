import React, {
  createContext,
  ReactElement,
  ReactNode,
  Reducer,
  useReducer,
} from "react";

import todoListReducer, { State } from "./reducers/todoListReducer";
import todoListActions, { Action } from "./actions/todoListActions";

type ContextType = {
  state: State;
  actions: Record<string, () => Promise<void>>;
};

type TodoListProviderProps = {
  children: ReactNode;
};

const initialState = { error: false, loading: false, data: [] };

const TodoListContext = createContext<ContextType>({
  state: initialState,
  actions: {},
});

const TodoAppProvider = ({ children }: TodoListProviderProps): ReactElement => {
  const [state, dispatch] = useReducer<Reducer<State, Action>>(
    todoListReducer,
    initialState
  );
  const actions = todoListActions(dispatch);

  return (
    <TodoListContext.Provider value={{ state, actions }}>
      {children}
    </TodoListContext.Provider>
  );
};

export { TodoListContext };
export default TodoAppProvider;
