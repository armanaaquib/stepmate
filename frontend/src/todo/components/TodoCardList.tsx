import React, { useContext, useEffect } from "react";

import TodoCard from "./TodoCard";
import StyledTodoCardList from "./style/TodoCardList.style";
import { TodoListContext } from "../context/TodoListProvider";

const TodoCardList = (): React.ReactElement => {
  const {
    state: { error, data },
    actions,
  } = useContext(TodoListContext);

  useEffect(() => {
    actions.loadTodoList();
  }, []);

  if (error) {
    return <div>Error in loading todo cards, Please refreseh the page</div>;
  }

  return (
    <StyledTodoCardList data-testid="todo-list">
      {data.map((todo) => (
        <TodoCard key={todo.id} entry={todo} />
      ))}
    </StyledTodoCardList>
  );
};

export default TodoCardList;
