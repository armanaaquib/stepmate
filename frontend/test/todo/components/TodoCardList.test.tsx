import React from "react";
import { render, RenderResult, screen } from "@testing-library/react";

import TodoCardList from "../../../src/todo/components/TodoCardList";
import { TodoListContext } from "../../../src/todo/context/TodoListProvider";
import {
  State,
  Todo,
} from "../../../src/todo/context/reducers/todoListReducer";

describe("<TodoCardList />", () => {
  const renderTodoCardList = (
    state: State,
    actions: Record<string, () => Promise<void>>
  ): RenderResult => {
    return render(
      <TodoListContext.Provider value={{ state, actions }}>
        <TodoCardList />
      </TodoListContext.Provider>
    );
  };

  it("should call loadTodoList action once", () => {
    const mockLoadTodoList = jest.fn();

    renderTodoCardList(
      { error: false, loading: false, data: [] },
      { loadTodoList: mockLoadTodoList }
    );

    expect(mockLoadTodoList).toHaveBeenCalledTimes(1);
  });

  it("should show error message if there is an error in loading todo list", () => {
    renderTodoCardList(
      { error: true, loading: false, data: [] },
      { loadTodoList: jest.fn() }
    );

    expect(screen.getByText(/error/i)).toBeInTheDocument();
  });

  it("should show todo cards", () => {
    const data: Todo[] = [
      {
        id: 1,
        title: "todo header",
        createdAt: "",
        modifiedAt: "",
        userId: 1,
        tasks: [
          {
            id: 1,
            text: "task",
            status: "IN_PROGRESS",
            createdAt: "",
            modifiedAt: "",
          },
        ],
      },
      {
        id: 2,
        title: "todo header",
        createdAt: "",
        modifiedAt: "",
        userId: 1,
        tasks: [],
      },
    ];

    const renderResult = renderTodoCardList(
      { error: false, loading: false, data },
      { loadTodoList: jest.fn() }
    );

    expect(renderResult).toMatchSnapshot();
  });
});
