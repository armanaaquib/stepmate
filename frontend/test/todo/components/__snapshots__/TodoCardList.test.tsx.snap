// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoCardList /> should show todo cards 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c5 {
  font-family: "Montserrat",sans-serif;
  font-size: 16px;
  font-weight: 400;
  margin-bottom: 8px;
  background-color: #ffffff;
  color: #000000;
  border-left: 4px solid #EEA940;
  border-radius: 2px;
  padding: 4px;
}

.c1 {
  min-height: 80px;
  min-width: 200px;
  max-width: 360px;
  padding: 12px 8px 12px 12px;
  margin: 16px;
  background-color: rgba(0,188,205,0.2);
  border-radius: 8px;
  box-shadow: 0 0 4px 2px rgba(0,188,205,0.3);
}

.c2 {
  padding-right: 56px;
}

.c3 {
  font-size: 20px;
  font-family: "Montserrat",sans-serif;
  font-weight: 500;
}

.c4 {
  padding-top: 16px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  padding: 52px;
}

<body>
    <div>
      <div
        class="c0"
        data-testid="todo-list"
      >
        <div
          class="c1"
        >
          <div
            class="c2"
          >
            <div
              class="c3"
            >
              todo header
            </div>
          </div>
          <div
            class="c4"
          >
            <div
              class=""
              data-testid="tasks"
            >
              <div
                class="c5"
              >
                task
              </div>
            </div>
          </div>
        </div>
        <div
          class="c1"
        >
          <div
            class="c2"
          >
            <div
              class="c3"
            >
              todo header
            </div>
          </div>
          <div
            class="c4"
          >
            <div
              class=""
              data-testid="tasks"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-iBPRYJ dooZzx"
      data-testid="todo-list"
    >
      <div
        class="sc-hKgILt fjRYYT"
      >
        <div
          class="sc-eCssSg dRQaNU"
        >
          <div
            class="sc-jSgupP ciFcwG"
          >
            todo header
          </div>
        </div>
        <div
          class="sc-gKsewC corgIH"
        >
          <div
            class="sc-gsTCUz"
            data-testid="tasks"
          >
            <div
              class="sc-bdfBwQ jLryXN"
            >
              task
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-hKgILt fjRYYT"
      >
        <div
          class="sc-eCssSg dRQaNU"
        >
          <div
            class="sc-jSgupP ciFcwG"
          >
            todo header
          </div>
        </div>
        <div
          class="sc-gKsewC corgIH"
        >
          <div
            class="sc-gsTCUz"
            data-testid="tasks"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
